### PR TITLE
Git-plugin: Add a "describe style" giving a relative commit name

### PR DESCRIPTION
**Pull request by ColinHebert**
Add a way to get a branch name with approximation.

With that, you can get branch names such as master~3 or origin^2~2 instead of a commit number.

### CODE REVIEW

**Comment by unphased** [comment]
I often need to refer the git commit hash (and a friendly relative branch/tag name, although very friendly, is too fuzzy/unclear). I've been trying to grok `prompt_sorin_setup` for a while, but making very little progress. I cannot tell how to modify this to force the `RPROMPT` to always show the commit hash.

**Comment by belak** [comment]
@unphased Yeah, it's a little convoluted. I'll try to give an overview here:

This portion goes over the formatting for git-info:
```
  # Set git-info parameters.
  zstyle ':prezto:module:git:info' verbose 'yes'
  zstyle ':prezto:module:git:info:action' format '%F{7}:%f%%B%F{9}%s%f%%b'
  zstyle ':prezto:module:git:info:added' format ' %%B%F{2}✚%f%%b'
  zstyle ':prezto:module:git:info:ahead' format ' %%B%F{13}⬆%f%%b'
  zstyle ':prezto:module:git:info:behind' format ' %%B%F{13}⬇%f%%b'
  zstyle ':prezto:module:git:info:branch' format ' %%B%F{2}%b%f%%b'
  zstyle ':prezto:module:git:info:commit' format ' %%B%F{3}%.7c%f%%b'
  zstyle ':prezto:module:git:info:deleted' format ' %%B%F{1}✖%f%%b'
  zstyle ':prezto:module:git:info:modified' format ' %%B%F{4}✱%f%%b'
  zstyle ':prezto:module:git:info:position' format ' %%B%F{13}%p%f%%b'
  zstyle ':prezto:module:git:info:renamed' format ' %%B%F{5}➜%f%%b'
  zstyle ':prezto:module:git:info:stashed' format ' %%B%F{6}✭%f%%b'
  zstyle ':prezto:module:git:info:unmerged' format ' %%B%F{3}═%f%%b'
  zstyle ':prezto:module:git:info:untracked' format ' %%B%F{7}◼%f%%b'
  zstyle ':prezto:module:git:info:keys' format \
    'status' '%b %p %c:%s%A%B%S%a%d%m%r%U%u'
```

The bottom-most setting is the overall display:
```
  zstyle ':prezto:module:git:info:keys' format \
    'status' '%b %p %c:%s%A%B%S%a%d%m%r%U%u'
```

Each of those keys can be looked up in https://github.com/sorin-ionescu/prezto/blob/master/modules/git/functions/git-info#L408 where we use this format string to include the actual data. The important thing to note is the 3 space-separated items at the start of the info format (followed by a :). These correspond to branch, position, and commit respectively.

We run git-info in an async task as seen here and use it to print out the overall status:
```
function prompt_sorin_async_git {
  cd -q "$1"
  if (( $+functions[git-info] )); then
    git-info
    print ${git_info[status]}
  fi
}
```

Finally, we parse this in the following by splitting the output into two fields on the `:`:
```
function prompt_sorin_async_callback {
  case $1 in
    prompt_sorin_async_git)
      # We can safely split on ':' because it isn't allowed in ref names.
      IFS=':' read _git_target _git_post_target <<<"$3"

      # The target actually contains 3 space separated possibilities, so we need to
      # make sure we grab the first one.
      _git_target=$(coalesce ${(@)${(z)_git_target}})

      if [[ -z "$_git_target" ]]; then
        # No git target detected, flush the git fragment and redisplay the prompt.
        if [[ -n "$_prompt_sorin_git" ]]; then
          _prompt_sorin_git=''
          zle && zle reset-prompt
        fi
      else
        # Git target detected, update the git fragment and redisplay the prompt.
        _prompt_sorin_git="${_git_target}${_git_post_target}"
        zle && zle reset-prompt
      fi
      ;;
  esac
}
```

There's some additional magic where we use coalesce which takes space separated arguments (which were the branch name, position, and commit) and returns the first one that's not empty:
```
      # The target actually contains 3 space separated possibilities, so we need to
      # make sure we grab the first one.
      _git_target=$(coalesce ${(@)${(z)_git_target}})
```

So, long story short:

* Remove the line which calls coalesce since you only want to return one value.
* Change the info format to the following (removing `%b %p` from the first part):
  ```
    zstyle ':prezto:module:git:info:keys' format \
      'status' '%c:%s%A%B%S%a%d%m%r%U%u'
  ```

That was probably more than you ever wanted to know, but I hope it helps.

**Comment by unphased** [comment]
Omg I love you

**Comment by diraol** [comment]
@belak this should go to a file and/or a page!!! Pretty awesome explanation!!!!

**Comment by unphased** [comment]
@belak Thank you for the explanation. I was able to use it to very quickly implement what I wanted, which was to move the `%c` from before the colon to after it, and it looks perfect.
